gazebo/Server.cc

```diff
@@ -542,6 +548,110 @@ void Server::Fini()
   gazebo::shutdown();
 }
 
+std::map<std::string, gazebo::common::Time> sensorsLastMeasurementTime;
+std::map<std::string, gazebo::common::Time> worldLastMeasurementTime;
+struct sensorPerformanceMetricsType
+{
+  double sensorRealUpdateRate;
+  double sensorSimUpdateRate;
+  double sensorFPS;
+};
+std::map<std::string, struct sensorPerformanceMetricsType> sensorPerformanceMetrics;
+/// \brief Publisher for run-time simulation performance metrics..
+transport::PublisherPtr performanceMetricsPub;
+
+void PublishPerformanceMetrics()
+{
+  physics::WorldPtr world = physics::get_world("default");
+
+  /// Outgoing run-time simulation performance metrics.
+  msgs::PerformanceMetrics performanceMetricsMsg;
+
+  // Real time factor
+  common::Time realTime = world->RealTime();
+  common::Time simTime = world->SimTime();
+  if (realTime == 0)
+    simTime = 0;
+  else
+    simTime = simTime / realTime;
+
+  if (simTime > 0)
+    performanceMetricsMsg.set_real_time_factor(simTime.Double());
+  else
+    performanceMetricsMsg.set_real_time_factor(0.0);
+
+  /// update sim time for sensors
+  for (auto model: world->Models())
+  {
+    for (auto link: model->GetLinks())
+    {
+      for (unsigned int i = 0; i < link->GetSensorCount(); i++)
+      {
+        std::string name = link->GetSensorName(i);
+        sensors::SensorPtr sensor = sensors::get_sensor(name);
+
+        auto ret = sensorsLastMeasurementTime.insert(
+            std::pair<std::string, gazebo::common::Time>(name, 0));
+        worldLastMeasurementTime.insert(
+                std::pair<std::string, gazebo::common::Time>(name, 0));
+        if (ret.second==false)
+        {
+          double updateSimRate =
+            (sensor->LastMeasurementTime() - sensorsLastMeasurementTime[name]).Double();
+          double updateRealRate =
+            (world->RealTime() - worldLastMeasurementTime[name]).Double();
+          sensorsLastMeasurementTime[name] = sensor->LastMeasurementTime();
+          if (updateSimRate > 0.0)
+          {
+            struct sensorPerformanceMetricsType emptySensorPerfomanceMetrics;
+            auto ret2 = sensorPerformanceMetrics.insert(
+                std::pair<std::string, struct sensorPerformanceMetricsType>
+                  (name, emptySensorPerfomanceMetrics));
+            if (ret2.second==false)
+            {
+              sensorPerformanceMetrics[name].sensorSimUpdateRate = 1.0/updateSimRate;
+              sensorPerformanceMetrics[name].sensorRealUpdateRate = 1.0/updateRealRate;
+              worldLastMeasurementTime[name] = world->RealTime();
+
+              // Special case for stereo cameras
+              sensors::CameraSensorPtr cameraSensor =
+                std::dynamic_pointer_cast<sensors::CameraSensor>(sensor);
+              if (nullptr != cameraSensor)
+              {
+                sensorPerformanceMetrics[name].sensorFPS = cameraSensor->Camera()->AvgFPS();
```

Comment:
The `AvgFPS` data came from ogre and I think it is the **real** time sensor update rate but averaged over a certain window size.

in this case, do we need to publish both the avg and instantaneous `real time sensor update rate`?

Reply:
The real time factor that is visualized in the Gazebo interface it's using a window of 20 values. The values that I'm publishing are instantaneous. 

is it fine to publish instantaneous values or should I create a small window?